ethclient/simulated/backend.go

```diff
@@ -0,0 +1,190 @@
+// from the Client interface returned by Backend.
+type simClient struct {
+	*ethclient.Client
+}
+
+// Client exposes the methods provided by the Ethereum RPC client.
+type Client interface {
+	ethereum.BlockNumberReader
+	ethereum.ChainReader
+	ethereum.ChainStateReader
+	ethereum.ContractCaller
+	ethereum.GasEstimator
+	ethereum.GasPricer
+	ethereum.GasPricer1559
+	ethereum.LogFilterer
+	ethereum.PendingStateReader
+	ethereum.PendingContractCaller
+	ethereum.TransactionReader
+	ethereum.TransactionSender
+	ethereum.ChainIDReader
+}
+
+// New creates a new binding backend using a simulated blockchain
+// for testing purposes.
+// A simulated backend always uses chainID 1337.
+func New(alloc core.GenesisAlloc, gasLimit uint64) *Backend {
+	// Setup the node object
+	nodeConf := node.DefaultConfig
+	nodeConf.DataDir = ""
+	nodeConf.P2P = p2p.Config{NoDiscovery: true}
+	stack, err := node.New(&nodeConf)
+	if err != nil {
+		// This should never happen, if it does, please open an issue
+		panic(err)
+	}
+
+	// Setup ethereum
+	genesis := core.Genesis{
+		Config:   params.AllDevChainProtocolChanges,
+		GasLimit: gasLimit,
+		Alloc:    alloc,
+	}
+	conf := ethconfig.Defaults
+	conf.Genesis = &genesis
+	conf.SyncMode = downloader.FullSync
+	conf.TxPool.NoLocals = true
+	sim, err := newWithNode(stack, &conf, math.MaxUint64)
+	if err != nil {
+		// This should never happen, if it does, please open an issue
+		panic(err)
+	}
+	return sim
+}
+
+// newWithNode sets up a simulated backend on an existing node
+// this allows users to do persistent simulations.
+// The provided node must not be started and will be started by newWithNode
+func newWithNode(stack *node.Node, conf *eth.Config, blockPeriod uint64) (*Backend, error) {
+	backend, err := eth.New(stack, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Register the filter system
+	filterSystem := filters.NewFilterSystem(backend.APIBackend, filters.Config{})
+	stack.RegisterAPIs([]rpc.API{{
+		Namespace: "eth",
+		Service:   filters.NewFilterAPI(filterSystem, false),
+	}})
+
+	// Start the node
+	if err := stack.Start(); err != nil {
+		return nil, err
+	}
+
+	// Set up the simulated beacon
+	beacon, err := catalyst.NewSimulatedBeacon(blockPeriod, backend)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reorg our chain back to genesis
+	if err := beacon.Fork(backend.BlockChain().GetCanonicalHash(0)); err != nil {
+		return nil, err
+	}
+
+	return &Backend{
+		eth:    backend,
+		beacon: beacon,
+		client: simClient{ethclient.NewClient(stack.Attach())},
+	}, nil
+}
+
+// Close shuts down the simBackend.
+// The simulated backend can't be used afterwards.
+func (n *Backend) Close() error {
+	if n.client.Client != nil {
+		n.client.Close()
+		n.client = simClient{}
+	}
+	if n.beacon != nil {
+		err := n.beacon.Stop()
+		n.beacon = nil
+		return err
+	}
+	return nil
+}
+
+// Commit seals a block and moves the chain forward to a new empty block.
+func (n *Backend) Commit() common.Hash {
```

Comment:
Is it weird at all to provide this to end users while also possibly having a non-zero (non-maxUint) period where blocks are sealing themselves? Would it make sense to require users of the simulated backend here to call commit when they want new blocks?

Reply:
This is needed for backward compatibility. The old sim backend only ever mined blocks on demand

Reply:
What I am proposing is to _only_ allow that functionality and not allow non-zero periods (where blocks are automatically sealed). If a caller wants that behavior they can roll a loop themselves.

Reply:
I don't fully understand. The non-zero periods are needed for the dev mode. The simulated backend will always set maxuint, so setting the period is not exposed to the user of the simbackend at all